### PR TITLE
feat: don't overcount the same field

### PR DESCRIPTION
**Pull request by jquense**
IDK if this is too simplistic, or if it is Good Enough (tm)

### CODE REVIEW

File: src/ComplexityVisitor.js [hunk 1]
```diff
@@ -1,8 +1,46 @@
...
+      ({ name }) => name.value === argument1.name.value,
+    );
+
+    if (!argument2) {
+      return false;
+    }
+
+    return print(argument1.value) === print(argument2.value);
+  });
+}
```
Comment (jquense) [inline src/ComplexityVisitor.js:23]: taken from upstream

File: src/ComplexityVisitor.js [hunk 2]
```diff
@@ -26,18 +64,33 @@ export default class ComplexityVisitor {
...
     this.Field = {
       enter: this.enterField,
       leave: this.leaveField,
     };
 
-    this.FragmentSpread = this.enterFragmentSpread;
+    this.FragmentDefinition = () => {
+      // don't visit any further we will include these at the spread location
+      return false;
+    };
```
Comment (jquense) [inline src/ComplexityVisitor.js:76]: logic for skipping and editing assumed to be correct as noted here: https://github.com/graphql/graphql-js/blob/master/src/language/visitor.js#L140

graphql-express and others do not use the edited AST so this should be safe

File: src/ComplexityVisitor.js [hunk 3]
```diff
@@ -1,8 +1,46 @@
...
+  });
+}
+
+function isSameSelection(selection1, selection2) {
+  return (
+    selection1.kind === selection2.kind &&
+    selection1.name?.value === selection2.name?.value &&
+    isSameArguments(selection1.arguments || [], selection2.arguments || [])
+  );
+}
```
Comment (jquense) [inline src/ComplexityVisitor.js:31]: should probably do something with directives?

File: src/CostCalculator.js [hunk 4]
```diff
@@ -10,28 +9,11 @@ export default class CostCalculator {
...
 
-  addFragment(costFactor, name) {
-    this.fragmentCosts.push([costFactor, name]);
-  }
-
-  calculateCost(fragmentCalculators) {
+  calculateCost() {
     if (this.cost !== null) {
       return this.cost;
     }
```
Comment (taion) [inline src/CostCalculator.js:15]: this stuff is probably irrelevant now

File: src/CostCalculator.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
 export default class CostCalculator {
```
Comment (taion) [inline src/CostCalculator.js:1]: which means this class is probably no longer necessary

File: src/ComplexityVisitor.js (hunk 1 above)
Comment (taion) [reply src/ComplexityVisitor.js:23]: would suggest leaving some comments re:code that's taken from elsewhere

File: src/ComplexityVisitor.js (hunk 3 above)
Comment (taion) [reply src/ComplexityVisitor.js:31]: directives (at least per relay) aren't supposed to affect execution in that way, so they shouldn't necessarily affect costs

File: src/ComplexityVisitor.js [hunk 6]
```diff
@@ -26,18 +64,33 @@ export default class ComplexityVisitor {
...
     this.rootCalculator = new CostCalculator();
-    this.fragmentCalculators = Object.create(null);
 
     this.Field = {
       enter: this.enterField,
       leave: this.leaveField,
     };
 
-    this.FragmentSpread = this.enterFragmentSpread;
+    this.FragmentDefinition = () => {
```
Comment (taion) [inline src/ComplexityVisitor.js:73]: i suspect there's a default "don't do anything" handler, as this doesn't handle all the AST types

File: src/CostCalculator.js (hunk 5 above)
Comment (jquense) [reply src/CostCalculator.js:1]: ya i know, i'll clean it up if we think the approach is good

File: src/ComplexityVisitor.js (hunk 6 above)
Comment (jquense) [reply src/ComplexityVisitor.js:73]: the comment/docs around return values is wrong for how visitor handlers work, `false` means `skip` vs `undefined` which means do nothing in particular. without the handler i believe it'll process the fragment selections

**Comment by taion** [comment]
btw this is a thing that should probably have test coverage

File: src/ComplexityVisitor.js (hunk 1 above)
Comment (itajaja) [reply src/ComplexityVisitor.js:23]: couldn't you do a deepEquals instead? if this is from upstream tho I guess that's ok

**Review by itajaja — COMMENTED** [review]
lgtm

**Comment by jquense** [comment]
idk why codecov isn't showing up here, it is working

**Pull request merged**